core/remote_state_verifier.go

```diff
@@ -0,0 +1,438 @@
+					verifyTaskFailedMeter.Mark(1)
+					task.Close()
+				}
+			}
+			vm.taskLock.Unlock()
+		case message := <-vm.messageCh:
+			vm.taskLock.RLock()
+			if vt, ok := vm.tasks[message.verifyResult.BlockHash]; ok {
+				vt.messageCh <- message
+			}
+			vm.taskLock.RUnlock()
+		// System stopped
+		case <-vm.bc.quit:
+			vm.taskLock.RLock()
+			for _, task := range vm.tasks {
+				task.Close()
+			}
+			vm.taskLock.RUnlock()
+			return
+		case <-vm.chainHeadSub.Err():
+			return
+		}
+	}
+}
+
+func (vm *remoteVerifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; header != nil && i <= maxForkHeight; i++ {
+		// if is genesis block, mark it as verified and break.
+		if header.Number.Uint64() == 0 {
+			vm.cacheBlockVerified(header.Hash())
+			break
+		}
+		func(hash common.Hash) {
+			// if verified cache record that this block has been verified, skip.
+			if _, ok := vm.verifiedCache.Get(hash); ok {
+				return
+			}
+			// if there already has a verify task for this block, skip.
+			vm.taskLock.RLock()
+			_, ok := vm.tasks[hash]
+			vm.taskLock.RUnlock()
+			if ok {
+				return
+			}
+
+			if header.TxHash == types.EmptyRootHash {
+				log.Debug("this is an empty block:", "block", hash, "number", header.Number)
+				vm.cacheBlockVerified(hash)
+				return
+			}
+
+			var diffLayer *types.DiffLayer
+			if cached, ok := vm.bc.diffLayerChanCache.Get(hash); ok {
+				diffLayerCh := cached.(chan struct{})
+				<-diffLayerCh
+				vm.bc.diffLayerChanCache.Remove(hash)
+				diffLayer = vm.bc.GetTrustedDiffLayer(hash)
+			}
+			// if this block has no diff, there is no need to verify it.
+			if diffLayer == nil {
+				log.Info("block's trusted diffLayer is nil", "hash", hash, "number", header.Number)
+				return
+			}
+			diffHash, err := CalculateDiffHash(diffLayer)
+			if err != nil {
+				log.Error("failed to get diff hash", "block", hash, "number", header.Number, "error", err)
+				return
+			}
+			verifyTask := NewVerifyTask(diffHash, header, vm.peers, vm.verifyCh, vm.allowInsecure)
+			vm.taskLock.Lock()
+			vm.tasks[hash] = verifyTask
+			vm.taskLock.Unlock()
+			verifyTaskCounter.Inc(1)
+		}(header.Hash())
+		header = vm.bc.GetHeaderByHash(header.ParentHash)
+	}
+}
+
+func (vm *remoteVerifyManager) cacheBlockVerified(hash common.Hash) {
+	if vm.verifiedCache.Len() >= verifiedCacheSize {
+		vm.verifiedCache.RemoveOldest()
+	}
+	vm.verifiedCache.Add(hash, true)
+}
+
+// AncestorVerified function check block has been verified or it's a empty block.
+func (vm *remoteVerifyManager) AncestorVerified(header *types.Header) bool {
+	// find header of H-11 block.
+	header = vm.bc.GetHeaderByNumber(header.Number.Uint64() - maxForkHeight)
+	// If start from genesis block, there has not a H-11 block.
+	if header == nil {
+		return true
+	}
+
+	hash := header.Hash()
+
+	// Check if the task is complete
+	vm.taskLock.RLock()
+	task, exist := vm.tasks[hash]
+	vm.taskLock.RUnlock()
+	if exist {
+		<-task.terminalCh
```

Comment:
It is a blocking operation here, I will suggest  wait 30 second and return false.

Reply:
> It is a blocking operation here, I will suggest wait 30 second and return false.

fixed

Reply:
1. `timeoutTicker` does not close which will cause a memory leak;
2.  `Ticker` is not a option here, please use `time.After`.
3.  `30` should be a constant variable.

Reply:
> 2\. a option here, please use

fixed